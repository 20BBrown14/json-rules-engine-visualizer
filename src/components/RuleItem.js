import React from 'react';
import PropTypes from 'prop-types';
import { Grid, IconButton } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import DeleteIcon from '@material-ui/icons/Delete';
import { SQFormTextField, SQFormDropdown, useSQFormContext } from '@selectquotelabs/sqform';
import { valueDropdownOptionsPropType } from '../util/proptypes';
import OPERATOR_SQFORMDROPDOWN_OPTIONS from '../constants/operatorConstants';

const itemGridStyles = makeStyles({
  containerGrid: {
    background: 'white',
    width: '100%',
    border: '1px solid #b3b3b3',
    marginBottom: '25px',
    height: '80px',
  },
  deleteIcon: {
    padding: 0,
  },
});

function RuleItem({
  ruleName,
  removeRuleItem,
  factNameDropdownOptions,
  valueDropdownOptions,
}) {
  const itemClasses = itemGridStyles();

  const { initialValues } = useSQFormContext();

  const [selectedFactName, setSelectedFactName] = React.useState(initialValues[`${ruleName}_factName`]);

  const handleFactNameDropdownChange = ({ target: { value: newFactName } }) => {
    setSelectedFactName(newFactName);
  };

  // Determine if SQForm has the values for these fields yet
  const initialValue = Object.keys(initialValues).find((key) => (
    `${ruleName}_factName` === key
      || `${ruleName}_operator` === key
      || `${ruleName}_value` === key
  ));

  const factField = React.useMemo(() => {
    if (factNameDropdownOptions) {
      return (
        <SQFormDropdown
          size={4}
          name={`${ruleName}_factName`}
          label="Fact Name"
          onChange={handleFactNameDropdownChange}
        >
          {factNameDropdownOptions}
        </SQFormDropdown>
      );
    }

    return (
      <SQFormTextField
        size={4}
        name={`${ruleName}_factName`}
        label="Fact Name"
        placeholder="Fact Name"
      />
    );
  }, [factNameDropdownOptions]);

  const valueField = React.useMemo(() => {
    if (Array.isArray(valueDropdownOptions)) {
      return (
        <SQFormDropdown
          size={4}
          name={`${ruleName}_value`}
          label="Value"
        >
          {valueDropdownOptions}
        </SQFormDropdown>
      );
    }

    // If is object
    if (valueDropdownOptions === Object(valueDropdownOptions) && factNameDropdownOptions) {
      if (!selectedFactName || valueDropdownOptions[selectedFactName]) {
        return (
          <SQFormDropdown
            size={4}
            name={`${ruleName}_value`}
            label="Value"
            isDisabled={!selectedFactName}
          >
            {valueDropdownOptions[selectedFactName] ?? []}
          </SQFormDropdown>
        );
      }
    }

    return (
      <SQFormTextField
        size={4}
        name={`${ruleName}_value`}
        label="Value"
        placeholder="Value"
      />
    );
  }, [valueDropdownOptions, factNameDropdownOptions, selectedFactName]);

  // If the initialValues don't exist don't render anything
  // Check needs to be here to ensure hooks are rendered in same order
  if (!initialValue) {
    return null;
  }

  return (
    <Grid container spacing={2} className={itemClasses.containerGrid}>
      <Grid item sm={1}>
        <IconButton className={itemClasses.deleteIcon} onClick={() => { removeRuleItem(ruleName); }}>
          <DeleteIcon />
        </IconButton>
      </Grid>
      {factField}
      <SQFormDropdown
        size={3}
        name={`${ruleName}_operator`}
        label="Operator"
      >
        {OPERATOR_SQFORMDROPDOWN_OPTIONS}
      </SQFormDropdown>
      {valueField}
    </Grid>
  );
}

RuleItem.propTypes = {
  /** Name of this particular rule */
  ruleName: PropTypes.string.isRequired,
  /** Function to delete rule */
  removeRuleItem: PropTypes.func.isRequired,
  /** Options to be used for the fact dropdown. */
  factNameDropdownOptions: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]).isRequired,
  })),
  /** Options to be used for the value dropdown */
  valueDropdownOptions: valueDropdownOptionsPropType,
};

export default RuleItem;
