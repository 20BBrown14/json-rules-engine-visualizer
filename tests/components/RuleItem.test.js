import React from 'react';
import RuleItem from '../../src/components/RuleItem';
import getRuleItemInitialValues from '../util/getInitialValues';
import SQFormWrapper from '../util/SQFormWrapper';

describe('RuleItem', () => {
  it('renders properly', () => {
    const testRuleItem = render(
      <SQFormWrapper initialValues={getRuleItemInitialValues('testRule')}>
        <RuleItem ruleName="testRule" />
      </SQFormWrapper>,
    );

    expect(testRuleItem).toMatchSnapshot();
  });
});
