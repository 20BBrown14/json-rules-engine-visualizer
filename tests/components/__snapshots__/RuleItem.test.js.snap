// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleItem renders properly 1`] = `
<form
  action="#"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      class="MuiGrid-root makeStyles-itemGrid-1 MuiGrid-container MuiGrid-spacing-xs-2"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
            data-shrink="true"
          >
            <span>
              Fact Name 
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input"
              name="testRule_factName"
              placeholder="Fact Name"
              type="text"
              value=""
            />
          </div>
          <p
            class="MuiFormHelperText-root"
          >
            <span>
              ​
            </span>
          </p>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
          id="operator"
        >
          Operator
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="operator mui-component-select-testRule_operator"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
            id="mui-component-select-testRule_operator"
            role="button"
            tabindex="0"
          >
            - -
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            name="testRule_operator"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
        <p
          class="MuiFormHelperText-root"
        >
          <span>
            ​
          </span>
        </p>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-4"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
            data-shrink="true"
          >
            <span>
              Value 
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input"
              name="testRule_valueName"
              placeholder="Value"
              type="text"
              value=""
            />
          </div>
          <p
            class="MuiFormHelperText-root"
          >
            <span>
              ​
            </span>
          </p>
        </div>
      </div>
    </div>
  </div>
</form>
`;
